Refactoring and 'if' expressions

jonsmock opened the pull request:
The primary goal of this PR is to setup adding 'expect' assertions, adding more interpolatable keys (e.g. 'working-directory'), and adding support for step outputs.  I wanted to refactor first, but the refactoring itself is a big enough change that I wanted to have it reviewed before building on top of it.  This also adds 'if' support, which I didn't intend to add, but it fell out pretty naturally.

Goals:

* Make a single place where assertions are described (currently just "exit code is zero")
* Collect stdout and stderr on a per retry basis, so we can assert about a single retry's stdout/stderr (not full history across retries)
* Make the step interpolation and execution easier to add to and easier to follow

To accomplish these goals, I've pushed the atom-based collection of stdout/stderr down into outer-spawn and docker-exec, having both return :stdout and :stderr keys, if we still want to collect the fully history across retries.  The outer-exec and compose-exec now have the same input and output interface, and they both now take the context, interpolated step, and opts (for logging).

execute-step* and execute-step are combined with the goal of colocating and simplifying that logic.  I think this is a step forward, but it's not quite where I'd like it to be.  The main tension is caused by how we skip steps.  On one hand, we want to skip _everything_, even step env interpolation, if we are skipping.  On the other hand, it would be very nice if the idea of skipping could be folded into the logic in more of a straight-line fashion, so that we're not constructing a result in two places of that big "if/else" in execute-step.  I was considering extracting an 'interpolate-step' function that implicitly stops interpolating things if the 'if' expression is false.  I deferred that, but if that sounds reasonable, I can apply that.  Btw, I'm also tempted to push down the 'verbose-commands' logging logic into outer-exec and compose-exec, as it's a bit distracting in execute-step.  Again open to opinions/suggestions.

Finally, this adds 'if' expression support to control executing/skipping of steps.  This is subtly different than GHA in two ways that are worth mentioning:

* our 'if' must take a bare expression (ex: `env.FOO == "BAR"`) and cannot be wrapped in `${{ ... }}`
* our 'if' expressions do not implicitly add `success()`, so to replicate the GHA behavior of `env.FOO == "BAR"`, a dctest user would have to write `success() && env.FOO == "BAR"`

I think we can support the GHA 'if' behavior, if we want to change either of these decisions.  I definitely do not have strong opinion here, and in fact, I implemented what I thought you would prefer :-D (simplicity > convenience).

Other misc changes:

* Use `jq --exit-status` so our GHA workflow fails on the jq results checking steps (not sure how I missed this)
* Add 'shell' support to steps, because `hostname`/etc wasn't working on NixOS
* Changes the JSON results-file format slightly, if you were depending on it with `--verbose-results` (I am not)

(Side note:  I'm happy to squash commits.  I wanted each commit to be very clear in isolation, but I realize that may have led to more commits than we'd want.)

jonsmock committed:
Allow setting 'shell' with 'run' strings

When 'run' is a string, allow the user to pick which shell to spawn/exec
with.  No change in behavior, when 'run' is an array.

Enum is currently limited to 'sh' and 'bash'.

jonsmock committed:
Return stdout/stderr keys from spawn/exec calls

Before this commit, we would collect all stdout/stderr across retry
attempts into a single atom; however, when we support assertions on
stdout/stderr, we'll want those assertions to only apply to a single
invocation/call.

This moves the collection of stdout/stderr into spawn and exec
separately, and they both return a common set of keys:  code, stdout,
stderr, and (optionally, for spawn) signal.

jonsmock committed:
Consolidate spawn/exec exit code assertion

Spawn and exec separately check and throw for a non-zero exit code.
Instead, let's simply return the results of the spawn/exec call, and let
the higher level function decide how to handle it.

jonsmock committed:
Narrow scope of try/catch in execute-step*

Moving the scope of the try/catch to be specifically around the call to
run-exec (either a spawn or exec). We add a :error key with contains
either an unexpected thrown error or an error from an assertion
(currently only the 'exit code == 0' assertion).

This change is not a pure refactoring, as it changes the output to the
results file if --verbose-results is used:

Old --verbose-results for successful and failed steps:

```
{"result":{"code":0,"stdout":"","stderr":""},"outcome":"pass"}
{"error":"Error running command: \"/bin/false\"","outcome":"fail"}
```

New --verbose-results for successful and failed steps:

```
{"code":0,"stdout":"","stderr":"","outcome":"pass"}
{"code":1,"stdout":"","stderr":"","error":{"message":"Error running command: \"/bin/false\""},"outcome":"fail"}
```

jonsmock committed:
GHA: use jq --exit-status to check results

These jq expressions return 'true' or 'false', but jq itself always
returns a zero exit code.  Setting `jq --exit-status` returns the
expected 0 or 1 exit code for 'true' or 'false' respectively.

jonsmock committed:
Combine execute-step and execute-step*

Having these functions split makes it difficult to implement things like
interpolating the step name, and it spreads the knowledge of the shape
of step results output across two functions. Let's combine them to
enable further refactoring.

jonsmock committed:
Colocate step interpolation logic, use same interface for execs

At a high-level the step logic is:

1) check 'if' condition*, exit early if necessary
2) interpolate step env
3) interpolate rest of step keys using interpolated env
4) run exec on host or container

* 'if' not directly supported yet

In this commit, we are moving steps 1-3 closer to each other and make
new interpolatable keys (e.g. working-directory) straightforward to add.

Additionally, pass context and step directly to outer-exec and
compose-exec to give them a standard input interface (their outputs are
already the same now).

jonsmock committed:
Support step 'if' expressions

This works subtly differently than GitHub Actions 'if' in two ways:

1. Only accepts expressions not interpolated text using ${{...}} syntax
2. Does not attempt an implicit "success()" in front of user expressions

The consequnce of bullet 2 is that a user might write "env.FOO == 2" in
GHA, but they will need to write "success() && env.FOO == 2" in our
expression langauge.

kanaka commented on schema.yaml:
```diff
@@ -48,3 +48,4 @@ properties:
                   interval: { "$ref": "#/$defs/interval", default: "1s" }
                   retries:  { type: integer } # no default (infinite retries)
                 default: { retries: 0 } # do not retry, if repeat is missing
+              shell: { type: string, enum: ["sh", "bash"], default: "sh" }
```
I suspect at some point we'll want to support full path to shell command but this is sufficient for now IMO.

jonsmock replied:
Agreed - figured I'd start restrictive, with the same default behavior, and we could widen as we chose.

kanaka commented on src/dctest/core.cljs:
```diff
@@ -71,32 +85,43 @@ Options:
-                      data)))
-    data))
+                      result)))
+    result))
```
My original thinking (that I should have made clearer) in moving the stdout/stderr collection out of these functions was that these functions (outer-spawn, outer-exec, docker-exec) feel like they are fairly generic and could move out of dctest into more general utility namespace(s) at some point. Passing in the handlers rather that the streams makes these somewhat specific to the shape that dctest wants.

Doing the stdout/stderr/stream hanlding in `execute-step*` had the secondary benefit that the dctest specific stdout/stderr accumulation code wasn't duplicated. I think that's still my preference. If we want to move this code out of `execute-step*` into another intermediate function that calls outer-exec/compose-exec (to enable assertions), that would be fine, but I think we should keep the outer-spawn/outer-exec/docker-exec more generic and not dctest specific.

jonsmock replied:
Makes a lot of sense to me.  I extracted 'run-exec, which I think both starts to make 'execute-step a lot clearer _and_ has the benefits that you're describing about the original.

kanaka commented on src/dctest/core.cljs:
```diff
@@ -60,22 +60,14 @@ Options:
+                                     (merge {:code code
+                                             :stdout (S/join "" @stdout)
+                                             :stderr (S/join "" @stderr)}
+                                            (when signal {:signal signal}))))))]
```
I haven't been 100% consistent about this myself, but we probably want to normalize naming for spawn vs. exec functions. Spawn tends to be stream based and oriented towards async usage and partial responses. Whereas exec, while technically async, tends to be used in a cases where the intermediate state is of no interest and it accumulates results and returns them all at once when things are done. And exec tends to itself towards returning good results and exceptions for bad results. I think we could have generic spawn functions for outer execution and docker/compose execution. Maybe generic exec functions for those too (which could accumulate and return stdout/stderr as part of the result). Or maybe the exec versions would be the dctest specific wrappers. Not 100%. But that was sort of the direction I was originally heading.

jonsmock replied:
Would this mean that what is currently called 'docker-exec should actually be called 'docker-spawn?  (I think I'm following and like what you're saying.)

kanaka replied:
Yes, I think so. Or perhaps there should be a docker-spawn and a docker-exec split (especially if those become part of the more generic utilities).

kanaka commented on examples/00-intro.yaml:
```diff
@@ -10,6 +10,9 @@ tests:
         run: /bin/true
+      - exec: node1
+        if: failure()
+        run: /bin/false
```
I think the intention is that one a test is failed that it stays failed even if there is a step that has `if: failure()` but the step itself returns true. In other words, we probably want to test two failure() cases in a row with the second one having a successful result. And then a final success() case that should never run (since the initial failure state should persist and not be modified by a failure() case that happens to have a successful result.

jonsmock replied:
Hrm, I was trying to illustrate/test that an obviously failing command would be skipped by `if: failure()`, if we have not yet failed.  But yeah, this neither particularly illustrative nor is it a great test.

I didn't do what you suggested quite, but I think the commit I added for this at least explains these function more clearly.  Let me know if you want me to keep going.

kanaka reviewed (changes requested):
Overall I like the results and final behavior. I do have reservations about the initial spawn callback refactoring. As noted inline, I think we have a clean spawn vs exec distinction and spawn functions should take I/O streams and only return final state (exit code).

jonsmock committed:
Restore stream-based outer-spawn and docker-exec

Reverting to the original, more generic outer-spawn and docker-exec that
accept streams for stdout/stderr. (Keeping 'shell' opts support)

Extracting 'run-exec from 'execute-step to be the bridge between dctest
and spawn/exec details.  'run-exec catches any exceptions, so that we
still get stdout/stderr in verbose results for any failed step commands.

Removes run-asserts (for now) as the non-zero exit code check has been
restored in both outer-spawn and docker-exec.

jonsmock committed:
Clarify success()/failure() in examples

kanaka reviewed (approved):
Looks great.
